Use GT_IND instead of LoadVector

EgorBo opened the pull request:
I noticed a suboptimal codegen for `Guid` comparisons, lets see what diffs think about @SingleAccretion's suggestion to fix that.

Example:
```csharp
bool Test(Guid g) => g == Guid.Empty;
```

Codegen diff:
```diff
; Assembly listing for method Proga:Test(System.Guid):bool:this
G_M54240_IG01:              
-      4883EC28             sub      rsp, 40
+      4883EC18             sub      rsp, 24
       C5F877               vzeroupper 
G_M54240_IG02:              
       C5F91002             vmovupd  xmm0, xmmword ptr [rdx]
-      C5F911442418         vmovupd  xmmword ptr [rsp+18H], xmm0
-      C5F857C0             vxorps   xmm0, xmm0, xmm0
       C5F911442408         vmovupd  xmmword ptr [rsp+08H], xmm0
-      C5FA6F442418         vmovdqu  xmm0, xmmword ptr [rsp+18H]
-      C5F974442408         vpcmpeqb xmm0, xmm0, xmmword ptr [rsp+08H]
-      C5F9D7C0             vpmovmskb eax, xmm0
-      3DFFFF0000           cmp      eax, 0xFFFF
+      C5F910442408         vmovupd  xmm0, xmmword ptr [rsp+08H]
+      C4E27917C0           vptest   xmm0, xmm0
       0F94C0               sete     al
       0FB6C0               movzx    rax, al
G_M54240_IG03:              
-      4883C428             add      rsp, 40
+      4883C418             add      rsp, 24
       C3                   ret      
-; Total bytes of code 59
+; Total bytes of code 39
```

EgorBo committed:
Use IND instead of LoadVector

tannergooding commented:
We have several memory load/store kinds. Many of these (aligned, non-temporal, etc) have specialized behavior or semantics that cannot be trivially represented by or converted to `IND` today.

I think it would be overall better if we found the place that's handling only `GT_IND` today and updated it to more generally handle memory loads instead, particularly since we ultimately want many of the same optimizations for the other memory load kinds.

EgorBo commented:
> We have several memory load/store kinds. Many of these (aligned, non-temporal, etc) have specialized behavior or semantics that cannot be trivially represented by or converted to `IND` today.
> 
> I think it would be overall better if we found the place that's handling only `GT_IND` today and updated it to more generally handle memory loads instead, particularly since we ultimately want many of the same optimizations for the other memory load kinds.

I agree that all loads should benefit from that but I think it's still better to have just one canonized version for all of them (except the multi-reg ones probably?) - `GT_IND` with corresponding flags instead of handling all kinds of HWIntrinsic loads everywhere in JIT (e.g. VN). Anyway, I'm leaving that up-for-grabs. This single-line change improved guid comparisons, removed more code than added, here is the diffs: https://dev.azure.com/dnceng-public/public/_build/results?buildId=67358&view=ms.vss-build-web.run-extensions-tab

EgorBo committed:
Update hwintrinsicxarch.cpp

SingleAccretion commented on src/coreclr/jit/hwintrinsicxarch.cpp:
```diff
@@ -1815,22 +1815,8 @@ GenTree* Compiler::impBaseIntrinsic(NamedIntrinsic        intrinsic,
-
-            retNode = gtNewSimdHWIntrinsicNode(retType, op1, loadIntrinsic, simdBaseJitType, simdSize);
+            retNode = gtNewIndir(retType, op1);
+            retNode->gtFlags |= GTF_IND_UNALIGNED;
```
Also needs `GTF_GLOB_REF`.

EgorBo committed:
Update hwintrinsicxarch.cpp

tannergooding commented:
> This single-line change improved guid comparisons, removed more code than added,

It also means that hardware intrinsic code may do less explicitly what users asked and we need to carefully validate that things are truly positive changes (including places where a different instruction is now used but the size hasn’t changed).

If we are doing this change, we should touch the other places that are using LoadVector128/256 as well. There are other import paths and places in lowering that create explicit hardware intrinsic nodes rather than IND nodes and it should be equally trivially to remove the non special hwintrinsic loads/stores so we can be consistent and get a better picture of the impact

EgorBo committed:
Cannonize more loads

EgorBo committed:
check ISAs

EgorBo committed:
Fix comments

EgorBo committed:
Handle stores

EgorBo committed:
Remove more code

EgorBo commented:
@tannergooding I moved all stores/loads to GT_IND except the aligned/nontemporal/paired ones
The PR is now `+33 −266` and diffs are much better: https://dev.azure.com/dnceng-public/public/_build/results?buildId=67848&view=ms.vss-build-web.run-extensions-tab

It also fixed addressing modes for stores e.g.:
```csharp
static void Test(ref byte x, nuint offset) => 
    Vector128<byte>.Zero.StoreUnsafe(ref x, offset);
```
Was:
```asm
8B010000          add     x0, x0, x1
4F000410          movi    v16.4s, #0
4C007010          st1     {v16.16b}, [x0]
```
Now:
```asm
4F000410          movi    v16.4s, #0
3CA16810          str     q16, [x0, x1]
```

> It also means that hardware intrinsic code may do less explicitly what users asked

Not sure I understand what you mean here, the initial version touched only crossplat helpers which don't promise any specific instructions. Anyway, we brake this invariant in many cases already with our optimizations/foldings for HWIntrinsic nodes.

tannergooding commented:
> Not sure I understand what you mean here, the initial version touched only crossplat helpers which don't promise any specific instructions. Anyway, we brake this invariant (which we never promised) in many cases already with our optimizations/foldings for HWIntrinsic nodes.

While we don't strictly guarantee any instruction, there is an entire experience at play here that needs to be considered for the HWIntrinsics and how they are expected to be used. I am just wanting to ensure we don't significantly regress things or make it harder for people who are optimizing their code because there are multiple factors to consider.

For example, in terms of disassembly readability even if `movups`/`movupd`/`movdqu` are equivalent in microcode, some of these are significantly easier to read/understands when taken with the context of surrounding code. Likewise, there are CPUs where these aren't equivalent and places where the type information is still important. AVX-512 w/ masking support brings back a distinction here as it impacts how the mask is interpreted and so knowing if its `byte` vs `word` vs `dword` vs `qword` becomes important once again.

To that end, I also don't want to see us do work here that we are going to revert and throw away in a couple months because it turns out that `GT_IND` can't support everything and so we ultimately need to support `HWIntrinsic` memory load operations anyways. -- I actually expect this will be the case specifically due to things like needing to track the base type and us not wanting to pay the additional cost required to support all the HWIntrinsic functionality for every `GT_IND`.

tannergooding commented on src/coreclr/jit/hwintrinsicarm64.cpp:
```diff
@@ -1242,123 +1242,28 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
-
         case NI_Vector64_LoadAlignedNonTemporal:
         case NI_Vector128_LoadAlignedNonTemporal:
-        {
```
nit: Please don't remove the braces. `FALLTHROUGH` is legal with them and they help with readability, browsability (you can collapse things in the IDE), and help clarify context so it avoids various issues that can crop up between case blocks.

tannergooding commented on src/coreclr/jit/hwintrinsicarm64.cpp:
```diff
@@ -1385,18 +1290,8 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
-
-            retNode = gtNewSimdHWIntrinsicNode(retType, op1, loadIntrinsic, simdBaseJitType, simdSize);
+            retNode = gtNewIndir(retType, op1);
+            retNode->gtFlags |= GTF_IND_UNALIGNED | GTF_GLOB_REF;
```
nit: Would be good to not set `GTF_IND_UNALIGNED` for the `LoadAligned*` intrinsics

tannergooding commented on src/coreclr/jit/hwintrinsicarm64.cpp:
```diff
@@ -1701,7 +1548,9 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
 
-            retNode = gtNewSimdHWIntrinsicNode(retType, op2, op1, NI_AdvSimd_Store, simdBaseJitType, simdSize);
+            retNode = gtNewOperNode(GT_IND, simdType, op2);
+            retNode->gtFlags |= (GTF_GLOB_REF | GTF_IND_UNALIGNED);
```
Same nit

SingleAccretion commented on src/coreclr/jit/hwintrinsicarm64.cpp:
```diff
@@ -1701,7 +1548,9 @@ GenTree* Compiler::impSpecialIntrinsic(NamedIntrinsic        intrinsic,
             }
 
-            retNode = gtNewSimdHWIntrinsicNode(retType, op2, op1, NI_AdvSimd_Store, simdBaseJitType, simdSize);
+            retNode = gtNewOperNode(GT_IND, simdType, op2);
```
Needs to be `gtNewIndir` or manually marked `GTF_EXCEPT`.

SingleAccretion commented on src/coreclr/jit/hwintrinsicxarch.cpp:
```diff
@@ -2282,22 +2213,9 @@ GenTree* Compiler::impBaseIntrinsic(NamedIntrinsic        intrinsic,
-            }
-
-            retNode = gtNewSimdHWIntrinsicNode(retType, op2, op1, storeIntrinsic, simdBaseJitType, simdSize);
+            retNode = gtNewOperNode(GT_IND, simdType, op2);
```
Needs to be `gtNewIndir` or manually marked `GTF_EXCEPT`.

EgorBo commented:
> For example, in terms of disassembly readability even if movups/movupd/movdqu are equivalent in microcode, some of these are significantly easier to read/understands when taken with the context of surrounding code. If Vector128.LoadVector128(float*) emits movups today, but changes to movdqu with this change, then it is a regression in the experience around HWIntrinsic usage.

It does not insert unaligned where previously it was aligned, I checked these 4:
```csharp
static Vector128<byte> Test1(ref byte x) =>
    Vector128.LoadUnsafe(ref x);

static Vector128<byte> Test2(byte* x) =>
    Vector128.Load(x);

static Vector128<byte> Test3(byte* x) =>
    Sse2.LoadVector128(x);

static Vector128<byte> Test4(ref byte x) =>
    Unsafe.ReadUnaligned<Vector128<byte>>(ref x);
```

> I actually expect this will be the case specifically due to things like needing to track the base type and us not wanting to pay the additional cost required to support all the HWIntrinsic functionality for every GT_IND.

What do you mean by the base type for a memory load? can't you get the relevant information from the address itself?

tannergooding commented:
> It does not insert unaligned where previously it was aligned, I checked these 4:

Those are a difference in type (float/double/integer) not in aligned vs unaligned. When reading disassembly, having a bunch of `ps` (float) instructions but then having a `dqu` (integer load/store) makes it harder to read and understand the logic.

> What do you mean by the base type for a memory load? can't you get the relevant information from the address itself? The additional cost will be to handle HW intrinsics everywhere where GT_IND currently works nice with SIMD (hence, the diffs) - e.g. valuenumber, lower (where addresses are marked as contained, hence, improved addressing mode on arm)

Not all memory operations are purely loads/stores from memory.

On most modern x86/x64 processors for SSE-AVX2 the instruction doesn't matter and so `movups` and `movdqu` can be freely interchanged. Both simply do 128-bit or 256-bit vector loads from memory. However, as per the above, using `movdqu` in the middle of a series of floating-point ops can make it harder to read/understand the code and so `Vector128<float> LoadVector128(float*)` should continue emitting `movups`. Likewise, there are some older and lower power processors where this difference can still matter (there can be different memory pipelines for float vs integer).

However, when considering AVX-512 and masking support (which we are adding for .NET 8), this difference between types actually becomes significant again as it impacts how the `kmask` register input is interpreted. Thus, AVX-512 introduced `vmovdqu8`, `vmovdqu16`, `vmovdqu32`, and `vmovdqu64` and the base type "must" be tracked when using the `EVEX` encoding.

I understand the immediate benefits of using `GT_IND`. My concern is that this is the overall wrong direction and we are doing work that will have to be reverted to support AVX-512. I'd expect we would be better off, overall, to generalize the `GT_IND` support as there are likely hwintrinsic support requirements that can never be added to `GT_IND` as it will make it too "expensive" for the common case.
